File: LeetCode/2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.cpp
int mod=1000000007;

class Solution {
private:
    vector<vector<vector<int>>> memo;
    
    int countPaths(vector<vector<int>>& grid,int i,int j,int s,int k,int n,int m)
    {
        if(i==n || j==m) 
            return 0;
        
        if(i==n-1 && j==m-1)
            return (s+grid[i][j])%k == 0;
        
        if(memo[i][j][s] != -1)
            return memo[i][j][s];
        
        return memo[i][j][s]=(countPaths(grid,i,j+1,(s+grid[i][j])%k,k,n,m) +
                           countPaths(grid,i+1,j,(s+grid[i][j])%k,k,n,m))%mod;
    }
    
public:
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        int n=grid.size();
        int m=grid[0].size();
        memo.resize(n+1,vector<vector<int>>(m+1,vector<int>(k+1,-1)));
        return countPaths(grid,0,0,0,k,n,m);
    }
};

File: LeetCode/2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.java
class Solution {
    public static int MOD = 1_000_000_007;

    private int numberOfPathsUtil(int i, int j, int remainder, int[][] grid, int k, int[][][] memo) {
        if(i == grid.length || j == grid[0].length) {
            return 0;
        }

        if(i == grid.length - 1 && j == grid[0].length - 1) {
            return (remainder + grid[i][j]) % k == 0 ? 1 : 0;
        }

        if(memo[i][j][remainder] != -1) {
            return memo[i][j][remainder];
        }

        int opt1 = numberOfPathsUtil(i, j + 1, (remainder + grid[i][j]) % k, grid, k, memo);
        int opt2 = numberOfPathsUtil(i + 1, j, (remainder + grid[i][j]) % k, grid, k, memo);
        
        return memo[i][j][remainder] = (opt1 + opt2) % MOD;
    }

    public int numberOfPaths(int[][] grid, int k) {
        int n = grid.length;
        int m = grid[0].length;
    
        int[][][] memo = new int[n][m][k];

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                Arrays.fill(memo[i][j], -1);
            }
        }

        return numberOfPathsUtil(0, 0, 0, grid, k, memo);
    }
}